examples/node/trx.c

```diff
@@ -0,0 +1,125 @@
+/* Copyright (c) 2019, Codership Oy. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "trx.h"
+
+#include <stdlib.h> // malloc()
+#include <errno.h>  // ENOMEM, etc.
+#include <stdbool.h>
+#include <assert.h>
+
+wsrep_status_t
+node_trx_execute(node_store_t* const   store,
+                 wsrep_t* const        wsrep,
+                 wsrep_conn_id_t const conn_id,
+                 size_t const          ws_size)
+{
+    wsrep_trx_id_t trx_id;
+    wsrep_key_t    ws_key;
+    wsrep_buf_t    ws = { NULL, ws_size };
+
+    /* prepare simple transaction and return the key and data buffer */
+    node_store_prepare_execute(store, &trx_id, &ws_key, &ws);
+    wsrep_status_t cert = WSREP_FATAL; // for cleanup
+
+    /* REPLICATION: get wsrep handle for transaction */
+    wsrep_ws_handle_t ws_handle = { trx_id, NULL };
+
+    /* REPLICATION: create a writeset for trx */
+    wsrep_status_t ret;
+    ret = wsrep->append_key(wsrep, &ws_handle,
+                            &ws_key, 1 /* single key */,
+                            WSREP_KEY_EXCLUSIVE, false /* no need to make a copy */);
+    if (ret) goto cleanup;
+    ret = wsrep->append_data(wsrep, &ws_handle,
+                             &ws, 1, WSREP_DATA_ORDERED, false);
+    if (ret) goto cleanup;
+
+    /* REPLICATION: replicate and certify the writeset */
+    static unsigned int const ws_flags =
+        WSREP_FLAG_TRX_START | WSREP_FLAG_TRX_END; // atomic trx
+    wsrep_trx_meta_t ws_meta;
+    cert = wsrep->certify(wsrep, conn_id, &ws_handle, ws_flags, &ws_meta);
+
+    /* REPLICATION: writeset was totally ordered, need to enter commit order */
+    if (ws_meta.gtid.seqno > 0)
```

Comment:
Special case of `WSREP_BF_ABORT` not handled here. Should at least be asserted/mentioned in comments.

Reply:
added `WSREP_BF_ABORT` check